### ISSUE TITLE
if (0.0 == 0.0) branch not eliminated in a complex shader

### ISSUE THREAD

**Issue by zeux** [KhronosGroup/SPIRV-Tools#1182]
In one of our shaders, the source contains compile-time dispatch using constants with `if` excluding the complex computation if necessary; SPIRV optimization pipeline we set up like this doesn't end up eliminating those:

```
spvtools::Optimizer tools(SPV_ENV_UNIVERSAL_1_0);

tools.RegisterSizePasses();
tools.RegisterPerformancePasses();
tools.Run(spirv_copy.data(), spirv_copy.size(), &spirv);
```

Attaching the offending SPIRV bytecode; an example of the branch in question is this one (fragment from spv-opt file):

       %1124 = OpFOrdEqual %bool %float_0 %float_0
               OpSelectionMerge %1125 None
               OpBranchConditional %1124 %1126 %1127

The branch does guard a texture sample, but since the branch condition is constant the optimizer should be able to constant-fold the condition to `true` and then eliminate one of the branches regardless of the texture sample instructions.

[shader-2NWLrkq.zip](https://github.com/KhronosGroup/SPIRV-Tools/files/1612440/shader-2NWLrkq.zip)

**Comment by dnovillo** [comment]
This is tracked in #1132

**Comment by zeux** [comment]
This is fixed by 105441360043248e91ec1c8b981f9541dab11083 and 04cd63e5b9c81b39906692cc635acfa2f4e1729f,
